File: plot/src/main/java/net/fwitz/math/plot/bifurc/LogisticMapPlot.java
package net.fwitz.math.plot.bifurc;

import javax.swing.*;
import java.util.function.Consumer;
import java.util.function.DoubleFunction;

import static java.util.Objects.requireNonNull;

public class LogisticMapPlot {
    private static final int DEFAULT_HEIGHT = 600;
    private static final int DEFAULT_WIDTH = 600;
    private static final double DEFAULT_MINR = 0.0;
    private static final double DEFAULT_MAXR = 4.0;
    private static final double DEFAULT_MINXN = 0.0;
    private static final double DEFAULT_MAXXN = 1.0;

    private final String title;

    protected DoubleFunction<double[]> computeFn;
    protected double minr = DEFAULT_MINR;
    protected double maxr = DEFAULT_MAXR;
    protected double minxn = DEFAULT_MINXN;
    protected double maxxn = DEFAULT_MAXXN;
    protected int width = DEFAULT_WIDTH;
    protected int height = DEFAULT_HEIGHT;

    private Consumer<LogisticMapPanel> panelDecorator = panel -> {
    };

    public LogisticMapPlot(String title) {
        this.title = requireNonNull(title, "title");
    }

    public final LogisticMapPlot domainR(double minr, double maxr) {
        this.minr = minr;
        this.maxr = maxr;
        return this;
    }

    public final LogisticMapPlot minr(double minr) {
        this.minr = minr;
        return this;
    }

    public final LogisticMapPlot maxr(double maxr) {
        this.maxr = maxr;
        return this;
    }

    public final LogisticMapPlot rangeXn(double minxn, double maxxn) {
        this.minxn = minxn;
        this.maxxn = maxxn;
        return this;
    }

    public final LogisticMapPlot minxn(double minxn) {
        this.minxn = minxn;
        return this;
    }

    public final LogisticMapPlot maxxn(double maxxn) {
        this.maxxn = maxxn;
        return this;
    }

    public final LogisticMapPlot width(int width) {
        if (width < 100 || width > 16384) {
            throw new IllegalArgumentException("width: " + width);
        }
        this.width = width;
        return this;
    }

    public final int width() {
        return width;
    }

    public final LogisticMapPlot height(int height) {
        if (height < 100 || height > 16384) {
            throw new IllegalArgumentException("height: " + height);
        }
        this.height = height;
        return this;
    }

    public final int height() {
        return height;
    }

    public final LogisticMapPlot size(int width, int height) {
        return width(width).height(height);
    }

    public final LogisticMapPlot computeFn(DoubleFunction<double[]> computeFn) {
        this.computeFn = requireNonNull(computeFn, "computeFn");
        return this;
    }

    public final LogisticMapPlot decoratePanel(Consumer<? super LogisticMapPanel> decorator) {
        this.panelDecorator = panelDecorator.andThen(decorator);
        return this;
    }

    public final void render() {
        final JFrame frame = new JFrame(title);
        frame.setSize(width(), height());
        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);

        final LogisticMapPanel panel = createPanel();
        frame.setContentPane(panel);

        panelDecorator.accept(panel);
        frame.setVisible(true);
    }

    protected LogisticMapPanel createPanel() {
        return new LogisticMapPanel(minr, minxn, maxr, maxxn, computeFn);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/bifurc/LogisticMapPlot.kt
package net.fwitz.math.plot.bifurc

import javax.swing.JFrame
import javax.swing.WindowConstants

class LogisticMapPlot(private val title: String) {
    companion object {
        private const val DEFAULT_HEIGHT = 600
        private const val DEFAULT_WIDTH = 600
        private const val DEFAULT_MINR = 0.0
        private const val DEFAULT_MAXR = 4.0
        private const val DEFAULT_MINXN = 0.0
        private const val DEFAULT_MAXXN = 1.0
    }

    protected var computeFn: ((Double) -> DoubleArray?)? = null
    protected var minr = DEFAULT_MINR
    protected var maxr = DEFAULT_MAXR
    protected var minxn = DEFAULT_MINXN
    protected var maxxn = DEFAULT_MAXXN
    protected var width = DEFAULT_WIDTH
    protected var height = DEFAULT_HEIGHT
    private var panelDecorator: (LogisticMapPanel) -> Unit = {}

    fun domainR(minr: Double, maxr: Double): LogisticMapPlot {
        this.minr = minr
        this.maxr = maxr
        return this
    }

    fun minr(minr: Double): LogisticMapPlot {
        this.minr = minr
        return this
    }

    fun maxr(maxr: Double): LogisticMapPlot {
        this.maxr = maxr
        return this
    }

    fun rangeXn(minxn: Double, maxxn: Double) = also {
        this.minxn = minxn
        this.maxxn = maxxn
    }

    fun minxn(minxn: Double): LogisticMapPlot = also { this.minxn = minxn }
    fun maxxn(maxxn: Double): LogisticMapPlot = also { this.maxxn = maxxn }

    fun width(width: Int): LogisticMapPlot = also {
        require(!(width < 100 || width > 16384)) { "width: $width" }
        this.width = width
    }

    fun height(height: Int): LogisticMapPlot {
        require(!(height < 100 || height > 16384)) { "height: $height" }
        this.height = height
        return this
    }

    fun size(width: Int, height: Int) = width(width).height(height)

    fun computeFn(computeFn: (Double) -> DoubleArray?): LogisticMapPlot {
        this.computeFn = computeFn
        return this
    }

    fun decoratePanel(decorator: (LogisticMapPanel) -> Unit): LogisticMapPlot {
        val prev = panelDecorator
        panelDecorator = { panel ->
            prev(panel)
            decorator(panel)
        }
        return this
    }

    fun render() {
        val frame = JFrame(title)
        frame.setSize(width, height)
        frame.defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        val panel = createPanel()
        frame.contentPane = panel
        panelDecorator(panel)
        frame.isVisible = true
    }

    protected fun createPanel(): LogisticMapPanel {
        return LogisticMapPanel(minr, minxn, maxr, maxxn, computeFn!!)
    }
}